app/src/main/java/com/example/taskyy/ui/viewmodels/LoginViewModel.kt

```diff
@@ -21,12 +24,20 @@ class LoginViewModel @Inject constructor(
     fun onEvent(event: LoginEvent) {
         when(event) {
             is LoginEvent.OnEmailChanged -> state = state.copy(email = event.email, isEmailValid = loginUseCase.isEmailValid(event.email))
-            is LoginEvent.OnLoginClick -> state = state.copy(isLoginSuccessful = loginUseCase.loginUser(event.login))
+            is LoginEvent.OnLoginClick -> login(event.login)
             is LoginEvent.OnPasswordChanged -> state = state.copy(password = event.password)
             is LoginEvent.OnTogglePasswordVisibility -> state = state.copy(isPasswordVisible = event.isPasswordVisible)
             is LoginEvent.OnNameChanged -> state = state.copy(name = event.name)
         }
     }
+
+    fun login(event: Login){
+        state = state.copy(isLogginIn = true)
+        viewModelScope.launch {
+            state = state.copy(isLoginSuccessful = loginUseCase.loginUser(event))
+        }
+        state = state.copy(isLogginIn = false)
```

Comment:
The first and last `state.copy` needs to go inside the `viewModelScope` block. The launched coroutine is running in parallel here independent of the "normal" execution flow. That means the code first updates `isLoggingIn` to true, then launches the coroutine (which happens instantly) and then updates `isLoggingIn` to false right after. But since the login call happens inside the coroutine, the function can't wait here until it finishes. That is only possible within the coroutine, so it would need to be like this:
```
viewModelScope.launch {
     state = state.copy(isLogginIn = true)
     state = state.copy(isLoginSuccessful = loginUseCase.loginUser(event))
     state = state.copy(isLogginIn = false) // Now this will be executed after loginUser
}
```